### ISSUE TITLE
Not possible to use step scope when allowBeanDefinitionOverriding is false [BATCH-2552]

### ISSUE THREAD

**Issue by spring-projects-issues** [spring-projects/spring-batch#1050]
**[Marcin Kiełbasiński](https://jira.spring.io/secure/ViewProfile.jspa?name=marcinkielbasinski@gmail.com)** opened **[BATCH-2552](https://jira.spring.io/browse/BATCH-2552?redirect=false)** and commented

Step scope is implemented as bean factory post processor and it registers proxy under the same name as target.

```
org.springframework.beans.factory.BeanDefinitionStoreException: Invalid bean definition with name 'TItemSupersessionReaderQueryProvider' defined in BeanDefinition defined in URL [file:/C:/workspaces/o
m/OM/om-app/target/jee-module-classes/META-INF/springbeans/custom/core/om-batch-beans.xml]: Cannot register bean definition [Root bean: class [org.springframework.aop.scope.ScopedProxyFactoryBean]; sc
ope=; abstract=false; lazyInit=false; autowireMode=0; dependencyCheck=0; autowireCandidate=true; primary=false; factoryBeanName=null; factoryMethodName=null; initMethodName=null; destroyMethodName=nul
l; defined in BeanDefinition defined in URL [file:/C:/workspaces/om/OM/om-app/target/jee-module-classes/META-INF/springbeans/custom/core/om-batch-beans.xml]] for bean 'TItemSupersessionReaderQueryProv
ider': There is already [Generic bean: class [com.xxx.om.batch.TItemSupersessionReaderQueryProvider]; scope=step; abstract=false; lazyInit=false; autowireMode=0; dependencyCheck=0; autowireCandida
te=false; primary=false; factoryBeanName=null; factoryMethodName=null; initMethodName=null; destroyMethodName=null; defined in URL [file:/C:/workspaces/om/OM/om-app/target/jee-module-classes/META-INF/
springbeans/custom/core/om-batch-beans.xml]] bound.
	at org.springframework.beans.factory.support.DefaultListableBeanFactory.registerBeanDefinition(DefaultListableBeanFactory.java:825)
	at org.springframework.batch.core.scope.BatchScopeSupport.createScopedProxy(BatchScopeSupport.java:159)
	at org.springframework.batch.core.scope.BatchScopeSupport.postProcessBeanFactory(BatchScopeSupport.java:130)
	at org.springframework.context.support.PostProcessorRegistrationDelegate.invokeBeanFactoryPostProcessors(PostProcessorRegistrationDelegate.java:284)
	at org.springframework.context.support.PostProcessorRegistrationDelegate.invokeBeanFactoryPostProcessors(PostProcessorRegistrationDelegate.java:174)
	at org.springframework.context.support.AbstractApplicationContext.invokeBeanFactoryPostProcessors(AbstractApplicationContext.java:674)
	at org.springframework.context.support.AbstractApplicationContext.refresh(AbstractApplicationContext.java:519)
...
```

---

**Affects:** 3.0.7

**Comment by mederel** [comment]
Hi,

affected by this problem too. Have to enable bean overrides to be able to use spring-batch :-(

I have written this simple maven project that allows to reproduce the issue: 
[spring-batch-issue-reproducer.tar.gz](https://github.com/spring-projects/spring-batch/files/4139539/spring-batch-issue-reproducer.tar.gz)

The issue seems to have been reported 4 years ago in 2016. I am not against working on fix, but I have a hard time understanding what can be done in the BatchScopeSupport and what is the use of each line. I would need help.

**Comment by mederel** [comment]
I tried to do a contribution with a fix... But I am having a hard time reproducing the issue with a unit test:

https://github.com/spring-projects/spring-batch/compare/4.2.x...mederel:bugfix/BATCH-2552-BeanDefinitionStoreException-with-StepScope?expand=1

The unit test I have modified raises a NoUniqueBeanDefinitionException, not a BeanDefinitionOverrideException...

**Comment by matrunich** [comment]
Hi all. 
Unfortunately - have the same issue. No any suggestions have not helped

Have app with several xml files with different batch job. 
Previously we  used WebSphere - now we are migrating on Spring Boot 2.

**Comment by mederel** [comment]
in definitive migrated successfully by using the @EnableBatchProcessing annotation, but with the modular = true option.

It allows to define a BatchConfigurer object that allowed to inject the customized beans (notably the PlatformTransactionManager).

Maybe that could be an option for you too?

BTW Thanks for the help in Gitter.

**Comment by matrunich** [comment]
I think current issue is:

https://docs.spring.io/spring-batch/docs/current/reference/html/step.html#step-scope

- Using a scope of Step is required in order to use late binding, because the bean cannot actually be instantiated until the Step starts, to allow the attributes to be found. Because it is not part of the Spring container by default, the scope must be added explicitly, b**y using the batch namespace or by including a bean definition explicitly for the StepScope, or by using the @EnableBatchProcessing annotation.** _Use only one of those methods._ The following example uses the batch namespace:
- And according to spring-batch-3.0.xsd 
	<xsd:annotation>
		<xsd:documentation><![CDATA[
	Defines the configuration elements for Spring Batch Core.
		]]></xsd:documentation>
	</xsd:annotation>

Sooo. In our case we have 
1) batch in namespace 
2) bean with the step
3) @ImportResours 

Have not checked yet but.  Bean factory tryes to declare beans by @ImportResours  
Then <batch:xxx> declares beans by it self. 


**Maybe that could be an option for you too?**
@mederel  Thanks for advice ! 

@mederel Does it work with allowBeanDefinitionOverriding =false ?

**Comment by fmbenhassine** [comment]
This should be resolved with #3942 . If anyone has the same issue with v5, please add a comment here and we will dig deeper.
